hoomd/mpcd/RejectionVirtualParticleFiller.h

```diff
@@ -0,0 +1,175 @@
+        {
+        m_exec_conf->msg->notice(5) << "Destroying MPCD RejectionVirtualParticleFiller" << std::endl;
+        }
+
+        //! Get the streaming geometry
+        std::shared_ptr<const Geometry> getGeometry() const
+            {
+            return m_geom;
+            }
+
+        //! Set the streaming geometry
+        void setGeometry(std::shared_ptr<const Geometry> geom)
+            {
+            m_geom = geom;
+            }
+
+        //! Fill the particles outside the confinement
+        void fill(unsigned int timestep);
+
+    protected:
+        std::shared_ptr<const Geometry> m_geom;
+
+    };
+
+
+template<class Geometry>
+void RejectionVirtualParticleFiller<Geometry>::fill(unsigned int timestep)
+    {
+    // Number of particles that we need to draw (constant)
+    const BoxDim& box = m_pdata->getBox();
+    const Scalar3 lo = box.getLo();
+    const Scalar3 hi = box.getHi();
+    const unsigned int m_NVirtMax = round(m_density*box.getVolume());
+
+    // Step 1: Create temporary GPUArrays to draw Particles locally using the worst case estimate for number
+    //         number of particles.
+    GPUArray<Scalar4> pos_loc(m_NVirtMax, m_exec_conf);
+    ArrayHandle<Scalar4> m_pos_loc(pos_loc, access_location::host, access_mode::readwrite);
+
+    // Step 2: Draw the particles.
+    unsigned int pidx = 0;
+    for (unsigned int i=0; i < m_NVirtMax; ++i)
+        {
+        // TODO: Currently just using the constant for SlitGeometryFiller which needs to be changed
+        // NOTE: particle tags neglected for the RNG here since we don't know them.
+        hoomd::RandomGenerator rng(hoomd::RNGIdentifier::SlitGeometryFiller, m_seed, timestep);
+
+        Scalar3 particle = make_scalar3(hoomd::UniformDistribution<Scalar>(lo.x, hi.x)(rng),
+                                       hoomd::UniformDistribution<Scalar>(lo.y, hi.y)(rng),
+                                       hoomd::UniformDistribution<Scalar>(lo.z, hi.z)(rng));
+
+        if (m_geom->isOutside(particle))
+            {
+            m_pos_loc.data[pidx] = make_scalar4(particle.x,
+                                              particle.y,
+                                              particle.z,
+                                              __int_as_scalar(m_type));
+            pidx += 1;
+            }
+        }
+
+    // in mpi, do a prefix scan on the tag offset in this range
+    // then shift the first tag by the current number of particles, which ensures a compact tag array
+    m_first_tag = 0;
+    #ifdef ENABLE_MPI
+    if (m_exec_conf->getNRanks() > 1)
+        {
+        // scan the number to fill to get the tag range I own
+        MPI_Exscan(&pidx, &m_first_tag, 1, MPI_UNSIGNED, MPI_SUM, m_exec_conf->getMPICommunicator());
+        }
+    #endif // ENABLE_MPI
+    m_first_tag += m_mpcd_pdata->getNGlobal() + m_mpcd_pdata->getNVirtualGlobal();
+
+    // Allocate memory for the new virtual particles.
+    m_mpcd_pdata->addVirtualParticles(pidx);
```

Comment:
It would be nice if this function returned the initial particle index (see math on line 132). You can modify it to do that then save the result like:
```suggestion
    const unsigned int first_idx = m_mpcd_pdata->addVirtualParticles(pidx);
```